File: frontend-app/src/app/services/car.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Car {
  id?: number;
  name: string;
  chassis: string;
  weight: number;
  acceleration: number;
  mpg: number;
  cost: number;
  formattedMpg?: string;
  formattedCost?: string;
  displayName?: string;
}

export interface CarStatistics {
  totalCars: number;
  averageMpg: number;
  averageCost: number;
  averageWeight: number;
  maxMpg: number;
  minCost: number;
  maxCost: number;
}

@Injectable({
  providedIn: 'root'
})
export class CarService {
  private baseUrl = 'http://localhost:8081/api/v1/cars';

  constructor(private http: HttpClient) {}

  getAllCars(): Observable<Car[]> {
    return this.http.get<Car[]>(this.baseUrl);
  }

  getCarById(id: number): Observable<Car> {
    return this.http.get<Car>(`${this.baseUrl}/${id}`);
  }

  createCar(car: Car): Observable<Car> {
    return this.http.post<Car>(this.baseUrl, car);
  }

  updateCar(id: number, car: Car): Observable<Car> {
    return this.http.put<Car>(`${this.baseUrl}/${id}`, car);
  }

  deleteCar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  searchCarsByName(name: string): Observable<Car[]> {
    const params = new HttpParams().set('name', name);
    return this.http.get<Car[]>(`${this.baseUrl}/search/name`, { params });
  }

  getCarsByChassis(chassis: string): Observable<Car[]> {
    const params = new HttpParams().set('chassis', chassis);
    return this.http.get<Car[]>(`${this.baseUrl}/filter/chassis`, { params });
  }

  getCarsByWeightRange(minWeight: number, maxWeight: number): Observable<Car[]> {
    const params = new HttpParams()
      .set('minWeight', minWeight.toString())
      .set('maxWeight', maxWeight.toString());
    return this.http.get<Car[]>(`${this.baseUrl}/filter/weight`, { params });
  }

  getFuelEfficientCars(minMpg: number): Observable<Car[]> {
    const params = new HttpParams().set('minMpg', minMpg.toString());
    return this.http.get<Car[]>(`${this.baseUrl}/filter/fuel-efficient`, { params });
  }

  getCarsInBudgetRange(minCost: number, maxCost: number): Observable<Car[]> {
    const params = new HttpParams()
      .set('minCost', minCost.toString())
      .set('maxCost', maxCost.toString());
    return this.http.get<Car[]>(`${this.baseUrl}/filter/budget`, { params });
  }

  searchCarsAdvanced(chassis?: string, minMpg?: number, maxCost?: number): Observable<Car[]> {
    let params = new HttpParams();
    if (chassis) params = params.set('chassis', chassis);
    if (minMpg) params = params.set('minMpg', minMpg.toString());
    if (maxCost) params = params.set('maxCost', maxCost.toString());
    
    return this.http.get<Car[]>(`${this.baseUrl}/search/advanced`, { params });
  }

  getAvailableChassisTypes(): Observable<string[]> {
    return this.http.get<string[]>(`${this.baseUrl}/chassis-types`);
  }

  getCarStatistics(): Observable<CarStatistics> {
    return this.http.get<CarStatistics>(`${this.baseUrl}/statistics`);
  }

  getMostFuelEfficientCars(limit: number = 5): Observable<Car[]> {
    const params = new HttpParams().set('limit', limit.toString());
    return this.http.get<Car[]>(`${this.baseUrl}/top-fuel-efficient`, { params });
  }
}

File: backend-api/src/main/java/com/icefaces/modernization/api/service/CarService.java
package com.icefaces.modernization.api.service;

import com.icefaces.modernization.api.dto.CarDto;
import com.icefaces.modernization.api.dto.CarStatisticsDto;
import java.util.List;
import java.util.Optional;

/**
 * Service interface for Car operations
 * Replaces business logic from ICEfaces managed beans like:
 * - org.icefaces.samples.showcase.example.compat.dataTable.DataTableBean
 * - org.icefaces.samples.showcase.example.ace.dataTable.DataTableBean
 * 
 * This interface defines the contract for car-related business operations
 * that were previously scattered across JSF managed beans.
 */
public interface CarService {
    
    /**
     * Retrieve all cars
     * Replaces: ICEfaces managed bean getData() methods
     */
    List<CarDto> getAllCars();
    
    /**
     * Find car by ID
     * Replaces: ICEfaces managed bean selection logic
     */
    Optional<CarDto> getCarById(Long id);
    
    /**
     * Create new car
     * Replaces: ICEfaces managed bean add/create methods
     */
    CarDto createCar(CarDto carDto);
    
    /**
     * Update existing car
     * Replaces: ICEfaces managed bean update/save methods
     */
    CarDto updateCar(Long id, CarDto carDto);
    
    /**
     * Delete car
     * Replaces: ICEfaces managed bean delete/remove methods
     */
    void deleteCar(Long id);
    
    /**
     * Search cars by name (case-insensitive)
     * Replaces: ICEfaces filter/search functionality
     */
    List<CarDto> searchCarsByName(String name);
    
    /**
     * Filter cars by chassis type
     * Replaces: ICEfaces dropdown filter logic
     */
    List<CarDto> getCarsByChassis(String chassis);
    
    /**
     * Find cars within weight range
     * Replaces: ICEfaces range filter components
     */
    List<CarDto> getCarsByWeightRange(Integer minWeight, Integer maxWeight);
    
    /**
     * Find fuel efficient cars (above specified MPG)
     * Replaces: ICEfaces business logic for filtering
     */
    List<CarDto> getFuelEfficientCars(Double minMpg);
    
    /**
     * Find cars within budget range
     * Replaces: ICEfaces cost filter functionality
     */
    List<CarDto> getCarsInBudgetRange(Double minCost, Double maxCost);
    
    /**
     * Advanced search with multiple criteria
     * Replaces: Complex ICEfaces filter combinations
     */
    List<CarDto> searchCars(String chassis, Double minMpg, Double maxCost);
    
    /**
     * Get available chassis types for dropdown population
     * Replaces: ICEfaces managed bean methods for populating select components
     */
    List<String> getAvailableChassisTypes();
    
    /**
     * Get car statistics for dashboard/reporting
     * Replaces: ICEfaces chart data generation logic
     */
    CarStatisticsDto getCarStatistics();
    
    /**
     * Get most fuel efficient cars (for recommendations)
     * Replaces: ICEfaces business logic for recommendations
     */
    List<CarDto> getMostFuelEfficientCars(int limit);
}